python/src/iceberg/table/metadata.py

```diff
@@ -0,0 +1,354 @@
+    remove oldest metadata log entries and keep a fixed-size log of the most 
+    recent entries after a commit."""
+
+    sort_orders: list
+    """A list of sort orders, stored as full sort order objects."""
+
+    default_sort_order_id: int
+    """Default sort order id of the table. Note that this could be used by 
+    writers, but is not used when reading because reads use the specs stored
+     in manifest files."""
+
+    def validate(self):
+        """Checks that the table metadata object is valid. The validation schema
+        used depends on the Iceberg table metadata version."""
+        if self.format_version == 1:
+            self.validate_v1(self.to_dict())
+        elif self.format_version == 2:
+            self.validate_v2(self.to_dict())
+        else:
+            raise ValueError(f"Unknown table metadata version {self.format_version}")
+
+    @staticmethod
+    def validate_v1(metadata: dict):
+        """Perform a JSONSchema validation using the v1 Iceberg table metadata schema"""
+        try:
+            validate_json(instance=metadata, schema=TABLE_METADATA_V1_SCHEMA)
+        except ValidationError as e:
+            # TODO Log something here
+            raise (e)
+
+    @staticmethod
+    def validate_v2(metadata: dict):
+        """Perform a JSONSchema validation using the v2 Iceberg table metadata schema"""
+        try:
+            validate_json(instance=metadata, schema=TABLE_METADATA_V2_SCHEMA)
+        except ValidationError as e:
+            # TODO Log something here
+            raise (e)
+
+    @classmethod
+    def from_byte_stream(cls, byte_stream, encoding="utf-8"):
+        """Instantiate a TableMetadata object from a byte stream
+
+        Args:
+            byte_stream: A file-like byte stream object
+            encoding (default "utf-8"): The byte encoder to use for the reader
+        """
+        reader = codecs.getreader(encoding)
+        metadata = json.load(reader(byte_stream))
+        return cls.from_dict(metadata)
+
+    @classmethod
+    def from_file(cls, path: str, custom_file_io: FileIO = None, **kwargs):
+        """Create a TableMetadata instance from a file path
+
+        Args:
+            path: A file path that can be read using an available FileIO implementation.
+            The FileIO implementation is determined using a string analysis on the path
+            by the iceberg.utils.file_io.get_file_io function
+            custom_file_io: A custom implementation of the iceberg.io.file.FileIO abstract
+            base class
+
+        Keyword Args:
+            Additional keyword arguments to pass forward to the instantiation of the appropriate
+            FileIO implementation.
+
+        """
+        if custom_file_io:
+            with custom_file_io(path, **kwargs) as f:
+                table_metadata = cls.from_byte_stream(byte_stream=f.byte_stream)
+        else:
+            with get_file_io(path, custom_file_io=custom_file_io, **kwargs) as f:
```

Comment:
it seems redundant to pass through custom_file_io here?  is the first branch really necessary?

Reply:
Great catch. The `get_file_io` function (not included in this PR) should be responsible for handling when a custom file-io is provided, i.e. `custom_file_io != None`. I'm planning to do another pass at this draft once PR #3691 is finalized and I'll make sure to update this. Thanks!